cedar-policy-core/src/ast/expr.rs

```diff
@@ -840,6 +840,307 @@ impl std::fmt::Display for Unknown {
+                    Expr::from(left.as_ref()),
+                    Expr::from(right.as_ref()),
+                )
+                .with_maybe_source_loc(source_loc)
+            }
+
+            proto::expr::expr_kind::Data::ExtApp(msg) => Expr::call_extension_fn(
+                Name::from(
+                    msg.fn_name
+                        .as_ref()
+                        .expect("as_ref() for field that will exist"),
+                ),
+                msg.args.iter().map(Expr::from).collect(),
+            )
+            .with_maybe_source_loc(source_loc),
+
+            proto::expr::expr_kind::Data::GetAttr(msg) => {
+                let arg = msg
+                    .expr
+                    .as_ref()
+                    .expect("as_ref() for field that will exist")
+                    .as_ref();
+                Expr::get_attr(Expr::from(arg), msg.attr.clone().into())
+                    .with_maybe_source_loc(source_loc)
+            }
+
+            proto::expr::expr_kind::Data::HasAttr(msg) => {
+                let arg = msg
+                    .expr
+                    .as_ref()
+                    .expect("as_ref() for field that will exist")
+                    .as_ref();
+                Expr::has_attr(Expr::from(arg), msg.attr.clone().into())
+                    .with_maybe_source_loc(source_loc)
+            }
+
+            proto::expr::expr_kind::Data::Like(msg) => {
+                let arg = msg
+                    .expr
+                    .as_ref()
+                    .expect("as_ref() for field that will exist")
+                    .as_ref();
+                Expr::like(Expr::from(arg), msg.pattern.iter().map(PatternElem::from))
+                    .with_maybe_source_loc(source_loc)
+            }
+
+            proto::expr::expr_kind::Data::Is(msg) => {
+                let arg = msg
+                    .expr
+                    .as_ref()
+                    .expect("as_ref() for field that will exist")
+                    .as_ref();
+                Expr::is_entity_type(
+                    Expr::from(arg),
+                    EntityType::from(
+                        msg.entity_type
+                            .as_ref()
+                            .expect("as_ref() for field that will exist"),
+                    ),
+                )
+                .with_maybe_source_loc(source_loc)
+            }
+
+            proto::expr::expr_kind::Data::Set(msg) => {
+                Expr::set(msg.elements.iter().map(Expr::from)).with_maybe_source_loc(source_loc)
+            }
+
+            proto::expr::expr_kind::Data::Record(msg) => Expr::record(
+                msg.items
+                    .iter()
+                    .map(|(key, value)| (key.into(), Expr::from(value))),
+            )
+            .expect("Expr should be valid")
+            .with_maybe_source_loc(source_loc),
+        }
+    }
+}
+
+#[cfg(feature = "protobufs")]
+impl From<&Expr> for proto::Expr {
```

Comment:
The obvious way to do this would require exposing a protobuf decode error upstream. I need to think more about the right approach, but I think this is ok for an experimental feature.